### ISSUE TITLE
CORS directive no works on Blazor Client App

### ISSUE THREAD

**Issue by Alejandro-Rufino** [aspnet/AspNetCore.Docs#16840]
[EDIT by guardrex to add the metadata. @Alejandro-Rufino, use the **This Page** feedback button+form at the bottom of any topic to open a new issue. No worries on this one: I just pasted the metadata into the bottom of this opening comment.]

I'm working in a test Blazor client app based on then MS-SQL Northwind database conected via Entity Framework.

In the  ConfigureServices method on the Startup.cs in my API  project I have:

```
public void ConfigureServices(IServiceCollection services)
        {
            services.AddControllers();
            services.AddDbContext<NorthWindDbContext>(option => option.UseSqlServer(Configuration.GetConnectionString("DefaultConnection")));

            Add Cors(Cross Origin Resource Sharing) so that the service can be accessed from other domains. 
            services.AddCors(o => o.AddPolicy("Policy", builder =>
            {
                builder.AllowAnyOrigin()
                  .AllowAnyMethod()
                  .AllowAnyHeader();
            }));
  }
```

In the Configure method:

```
public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
        {
            if (env.IsDevelopment())
            {
                app.UseDeveloperExceptionPage();
            }

            app.UseHttpsRedirection();

            app.UseRouting();

             app.UseEndpoints(endpoints =>
            {
                 endpoints.MapDefaultControllerRoute();
            });

            app.UseCors("Policy");
}
```
With this configuration (based on what I've read) the CORS policy should work, but it doesn't work.

In the controller class I had to add the following line: 
    Response.Headers.Add("Access-Control-Allow-Origin", "*");

The method like this:

```
 [HttpGet("{id}")]
 public async Task<ActionResult<Category>> GetCategory(int id)
 {            
           var category = await _context.Categories.FindAsync(id);

            if (category == null)
            {
                return NotFound();
            }

            Response.Headers.Add("Access-Control-Allow-Origin", "*");
            return category;
        }
```

With this line the GET method woks fine but the PUT, POST and DELETE Methods no works.

The PUT Method:

```
public async Task<IActionResult> PutCategory(int id, Category category)
        {
            Response.Headers.Add("Access-Control-Allow-Origin", "*");

            if (id != category.CategoryID)
            {
                return BadRequest();
            }

            _context.Entry(category).State = EntityState.Modified;

            try
            {
                await _context.SaveChangesAsync();
            }
            catch (DbUpdateConcurrencyException)
            {
                if (!CategoryExists(id))
                {
                    return NotFound();
                }
                else
                {
                    throw;
                }
            }
     
            return NoContent();
        }
```

In the client app the save method:

```
async Task SaveCategory(Category category)
    {
        string baseUrl = "http://localhost:60894/";
        string endpoint = $"{baseUrl}api/categories";
        url = endpoint;      

        try
        {          
            if (category.CategoryID != 0)
            {
                await httpClient.SendJsonAsync(HttpMethod.Put, endpoint + "/" + category.CategoryID, category);
            }
            else
            {
                await  httpClient.SendJsonAsync(HttpMethod.Post, endpoint, category);
            }     
        }
        catch (Exception e)
        {
            showError = true;
        }
    }
```

When i run the app, I have the same error messages:

blazor.webassembly.js:1 OPTIONS http://localhost:60894/api/categories/3 405 (Method Not Allowed)

and

Access to fetch at 'http://localhost:60894/api/categories/3' from origin 'http://localhost:63978' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.

This problem occurs with all browsers.

A try to solve the problem adding diferents options to UseCors in the API project and diferent headers in client app but nothing woks.

Anybody have any idea where the mistake is?

Regards.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 59bca5db-1196-2a8b-a0d8-4e89b2382288
* Version Independent ID: d3f332c4-4c60-039e-0e1e-bc6b11831f87
* Content: [Enable Cross-Origin Requests (CORS) in ASP.NET Core](https://docs.microsoft.com/en-us/aspnet/core/security/cors?view=aspnetcore-3.1#enable-cors-with-attributes)
* Content Source: [aspnetcore/security/cors.md](https://github.com/aspnet/AspNetCore.Docs/blob/master/aspnetcore/security/cors.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-security**
* GitHub Login: @Rick-Anderson
* Microsoft Alias: **riande**

**Comment by guardrex** [comment]
Hello @Alejandro-Rufino ... Not sure about some of the error info at the end yet; but first, move your call to `UseCors`. See the **Warning** at https://docs.microsoft.com/en-us/aspnet/core/security/cors?view=aspnetcore-3.1#apply-cors-policies-to-all-endpoints.

**Comment by guardrex** [comment]
I'm going to close as both (possibly) answered and as a support request. If you need support, try support channels ...

* [Stack Overflow](https://stackoverflow.com/questions)
* [ASP.NET Core Slack](http://tattoocoder.com/aspnet-slack-sign-up/)
* [ASP.NET Gitter](https://gitter.im/aspnet/Home)

If you find a problem with a doc, please use the **This Page** feedback button+form at the bottom of any topic.
